### ISSUE TITLE
Dynamic Template Literals can't be Mapped

### ISSUE THREAD

**Issue by MatthewAry** [sinclairzx81/typebox#913]
I'm trying to do some meta-programming for making APIs on Elysia. Using TS 5.5.2

Getting an interesting problem
```ts
function baseNumericComparableLiteral<T extends string>(key: T) {
  const comparables = Type.TemplateLiteral([Type.Union([
    Type.Literal('gt'),
    Type.Literal('gte'),
    Type.Literal('lt'),
    Type.Literal('lte'),
    Type.Literal('eq'),
    Type.Literal('ne'),
  ])])
  return Type.TemplateLiteral([Type.Literal(key), comparables])
}

const literals = baseNumericComparableLiteral('key')
type literalsType = Static<typeof literals> // type literalsType = "keygt" | "keygte" | "keylt" | "keylte" | "keyeq" | "keyne"

const mapping = Type.Mapped(literals, K => Type.Date())
type mappingType = Static<typeof mapping> // type mappingType = { key: Date; }
```

On `mappingType`, I expected the result to match to:
```ts
type mappingType = {
  "keygt": Date
  "keygte": Date
  "keylt": Date
  "keylte": Date
  "keyeq": Date
  "keyne": Date
}
```

Am I missing something?

**Comment by sinclairzx81** [comment]
@MatthewAry Hi, thanks for reporting!

> On mappingType, I expected the result to match to:

You can get this to work with the following.

[TypeScript Link Here](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzgFQJ5gKYBo4GUYCGMwAxnAL5wBmUEIcA5AAIDOwAdiQDYHBQD0MdBgBGEAB4MAUFKoBXTsQjs4IgiwwA5OSAxRSAYTpgCUAiK4YAMsBh6CXADzI4GcXfYATFnBYx97ADmAHwAFADWGKgAXCgAlIhScHD8-HAEAG4QwJ6uICIYnp4cgXB24Dx2cFy29lxwHGUYFUQY1bVmXElwUBgwclAqaJgAdMjNYJXWHQ6hANrDGCM2dp0RUXE4iyMAquzAyvNw3cnbK3WhDIEwDJsnKMLLM1yX1xi3WPdnz5dcN3fJU6Pc5rBh-d4AwHfVazBgYACOHy+wJ+DHYEM+yQAunEcVJyDISMo-O0YVwfABeVTqLQ6PSGYymcyWEGwyKoW5SISYUl1FiLOBU-BEUiObkYCBUXmdFjBFJpcXShz84SCuAAInZ13VcAAPhqtXYdfrNVE-saDWajXrLagERbTah0erCcT4CACGAwCU1dsALKezCeUI1MksHAAaUFcu2ABFWqE4nEuaqPV6SgKhYRiCQxcJJXA096gnLUolARXK1XqzXa2XbdrYvG7Pda232zX6463uqm61Wx3Bx2u+zzX2W0PJ4OR1aML24M2MAOpyvATO7fD54vl6uV+vneOl7vj2u0uQgA)

```typescript
function baseNumericComparableLiteral<T extends string>(key: T) {
  // avoid embedding template literal in template literal
  return Type.TemplateLiteral([Type.Literal(key), Type.Union([ 
    Type.Literal('gt'),
    Type.Literal('gte'),
    Type.Literal('lt'),
    Type.Literal('lte'),
    Type.Literal('eq'),
    Type.Literal('ne'),
  ])])
}

const literals = baseNumericComparableLiteral('key')
type literalsType = Static<typeof literals> // type literalsType = "keygt" | "keygte" | "keylt" | "keylte" | "keyeq" | "keyne"

const mapping = Type.Mapped(literals, K => Type.Date())
type mappingType = Static<typeof mapping> // {
                                          //  "keygt": Date
                                          //  "keygte": Date
                                          //  "keylt": Date
                                          //  "keylte": Date
                                          //  "keyeq": Date
                                          //  "keyne": Date
                                          // }
```
However I would expect the embedded template literal (as per your example), to correctly map also. I will take a look at this.

Let me know if the above helps
Cheers
S

**Comment by sinclairzx81** [comment]
@MatthewAry Hiya,

Have pushed a fix for this on 0.32.34. You should be able to embed template within templates as per your original example. The following should map as you expect (note: you may need to clear your browser cache)

[TypeScript Link Here](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzgFQJ5gKYBo4GUYCGMwAxnAL5wBmUEIcA5AAIDOwAdiQDYHBQD0MdBgBGEAB4MAUDKoBXTsQjs4IgiwwA5OSAxRSAYTpgCUAiK4YAMsBh6CXADzI4GcXfYATFnBYx97ADmAHwAFADWGKgAXCgAlIhScHAkyn4pxqbmlj4AvCjCAHTIGOA8djZ2ZlyhANpomIUAquzAynVJyQWNlfY1DIEwDHFYnckNGIW91aEDdsOjXd2T0w6zXEMjY8tTtn3r81tLE7tVawwYAI4L2yer-ewYN8kAunFvnVAYMHJQKiclMpEax7Gb1Ir3CJREYZcBZCwYFgfcgyVLsdJcUEOPKqdRaHR6QyZMwIyEMSKoYZSISYOCYs5cFgTOD5fBEUiOGkYCBUOlYxnBOD8fhwLl8hlM4QsuAAIgpgxlcAAPrL5XZFSq5VENhrVdr1cq9agrrqtahHjLUWl4CACGAwBxAtKTgBZO2YTyhel9Fg4ADSLMFJwAIsDQnE4tSpbb7Y7mazCMQSJzhDy4DGHUFBcLRdH3XGpfkkBTYqG7ABuChSIA)

This issue was specific to the inference, where the runtime mapping was working as expected. TypeBox is due to revisit the template literal implementation under the revised inference architecture (which was added on 0.32.x), there's are still known issues with embedding unions in unions in template literals in unions (or some nested permutation of) (which is technically possible in TS, but quite challenging to map), This said, supporting embedded template literals in template literals was envisioned for the first pass of this feature (so thanks for finding this)

If you run into any issues, feel free to ping on this thread.
Cheers
S
